### PR TITLE
Tabs scroll animation

### PR DESCRIPTION
**Pull request by HansMuller**
The TabBar's selection is now represented by a TabBarSelection object which encapsulates both the previous and currently selected indices and the Performance used to animate the selection indicator.

Added a TabBarView class which displays a tab's contents. It uses a shared TabBarSelection to stay in sync with a TabBar. The TabBarView scrolls in sync with the TabBar when the selection changes. Eventually it will allow one to fling the selection forward or backwards.

Added a tabBar property to ToolBar. Typically the corresponding TabBarView will be the body of the toolbar's Scaffold.

Removed TabNavigatorView and TabNavigator.

Added a widget gallery tabs demo page. Removed the old tabs demo.

### CODE REVIEW

File: examples/material_gallery/lib/demo/tabs_demo.dart [hunk 1]
```diff
@@ -0,0 +1,59 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+
+import 'widget_demo.dart';
+
+final TabBarSelection _selection = new TabBarSelection();
+final List<String> _iconNames = ["event", "home", "android", "alarm", "face", "language"];
```
Comment (abarth) [inline examples/material_gallery/lib/demo/tabs_demo.dart]: `<String>[ ... ]`

File: examples/material_gallery/lib/demo/tabs_demo.dart [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+Widget buildTabBar(_) {
+  return new TabBar(
+    selection: _selection,
+    isScrollable: true,
+    labels: _iconNames.map((String iconName) => new TabLabel(text: iconName, icon: "action/$iconName")).toList()
+  );
+}
```
Comment (abarth) [inline examples/material_gallery/lib/demo/tabs_demo.dart]: I'd made this a member function of _TabsDemoState and declare the parameter.  The idea behind the material_gallery is that it's sample code that people will copy/paste, which means we should try to make the code in it use patterns that we want people to follow.

File: examples/material_gallery/lib/demo/tabs_demo.dart [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  double _viewWidth = 100.0;
+
+  void _handleSizeChanged(Size newSize) {
+    setState(() {
+      _viewWidth = newSize.width;
+    });
+  }
+
+  Widget build(_) {
+    return new SizeObserver(
```
Comment (abarth) [inline examples/material_gallery/lib/demo/tabs_demo.dart]: Why do we need to have this size observer?  Can't the TabBarView make this itself?

File: examples/material_gallery/lib/demo/tabs_demo.dart [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        }
+      )
+    );
+  }
+}
+
+final WidgetDemo kTabsDemo = new WidgetDemo(
+  title: 'Tabs',
+  routeName: '/tabs',
+  tabBarBuilder: buildTabBar,
```
Comment (abarth) [inline examples/material_gallery/lib/demo/tabs_demo.dart]: Hum...  I'm worried where this path will lead, but I'm not sure what's better to suggest at the moment.

File: packages/flutter/lib/src/material/tabs.dart [hunk 5]
```diff
@@ -381,18 +377,46 @@ class _TabsScrollBehavior extends BoundedBehavior {
   }
 }
 
+class TabBarSelection {
+  TabBarSelection({ int index: 0, this.onChanged }) : _index = index;
+
+  VoidCallback onChanged;
```
Comment (abarth) [inline packages/flutter/lib/src/material/tabs.dart]: final ?

File: packages/flutter/lib/src/material/tabs.dart [hunk 6]
```diff
@@ -563,46 +615,111 @@ class _TabBarState extends ScrollableState<TabBar> {
...
+
+  TabBarViewState createState() => new TabBarViewState<T>();
+}
+
+// TODO(hansmuller): horizontal scrolling should drive the TabSelection's performance.
+class _NotScrollable extends BoundedBehavior {
+  bool get isScrollable => false;
+}
+
+class TabBarViewState<T> extends ScrollableListState<T, TabBarView<T>> {
```
Comment (abarth) [inline packages/flutter/lib/src/material/tabs.dart]: `s/TabBarViewState/_TabBarViewState/` ?

**Comment by abarth** [comment]
LGTM

File: examples/material_gallery/lib/demo/tabs_demo.dart (hunk 3 above)
Comment (HansMuller) [inline examples/material_gallery/lib/demo/tabs_demo.dart]: Yes, I was pretty unhappy having the SizeObserver be the responsibility of the caller.  It didn't report the correct size when I just had the TabBarView incorporate it directly.  I'll try again; sort out exactly what the problem is/was.

File: examples/material_gallery/lib/demo/tabs_demo.dart (hunk 4 above)
Comment (HansMuller) [inline examples/material_gallery/lib/demo/tabs_demo.dart]: I didn't like it either. I could give WidgetDemo a public buildToolBar() method and then create a subclass of WidgetDemo here.

File: packages/flutter/lib/src/material/tabs.dart (hunk 5 above)
Comment (HansMuller) [inline packages/flutter/lib/src/material/tabs.dart]: Yes!

File: packages/flutter/lib/src/material/tabs.dart (hunk 6 above)
Comment (HansMuller) [inline packages/flutter/lib/src/material/tabs.dart]: Good point, yes.

**Comment by HansMuller** [comment]
I'd forgotten to update the Stocks demo. Adding that now.

**Pull request merged**